### PR TITLE
adding tests to segmentation.py

### PR DESCRIPTION
**Pull request by Kamil-Sabbagh**

### CODE REVIEW

File: innofw/core/models/torch/lightning_modules/segmentation.py [hunk 1]
```diff
@@ -136,6 +136,7 @@ def stage_step(self, stage, batch, do_logging=False, *args, **kwargs):
 
         if stage in ["train", "val"]:
             loss = self.log_losses(stage, predictions, label)
+            self.log("loss", loss)
```
Comment (QazyBi) [inline innofw/core/models/torch/lightning_modules/segmentation.py:139]: not needed. As function above does the same but also adds stage prefix

**Review by QazyBi — CHANGES_REQUESTED** [review]

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        return DataLoader(self.dataset, batch_size=self.batch_size)
+
+    def val_dataloader(self):
+        return DataLoader(self.dataset, batch_size=self.batch_size)
+
+    def test_dataloader(self):
+        return DataLoader(self.dataset, batch_size=self.batch_size)
+
+@pytest.fixture
+def segmentation_module() -> LightningModule:
```
Comment (sham1lk) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:67]: let's put fixture to the fixture folder, or conftest.py file

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    dataset = DummyDataset(num_samples=100)
+    dataloader = DataLoader(dataset, batch_size=4)
+
+    # Fit the model
+    trainer.test(segmentation_module, dataloaders=dataloader)
+
+
+
+
+
```
Comment (sham1lk) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:139]: Are you using pre-commit? let's delete this commented code if it's not needed

Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py]: thanks for noting the style problem. I will be fixing it

File: innofw/core/models/torch/lightning_modules/segmentation.py (hunk 1 above)
Comment (Kamil-Sabbagh) [reply innofw/core/models/torch/lightning_modules/segmentation.py]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 2 above)
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:77]: Can you please the path to where can I add it? Just to make sure I will be adding it in the correct place.

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    trainer = Trainer(max_epochs=1, devices=1)
+    dataset = DummyDataset(num_samples=100)
+    dataloader = DataLoader(dataset, batch_size=4)
+    trainer.test(segmentation_module, dataloaders=dataloader)
+
+
+def test_predicting_without_checkpoint(segmentation_module: LightningModule):
+    trainer = Trainer(max_epochs=1, devices=1)
+    predict_dataloader = DataLoader(DummyDataset(10), batch_size=4)
+    predictions = predict(segmentation_module, predict_dataloader)
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:126]: use `trainer.predict` here

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+    # Fit the model (train and validate)
+    trainer.fit(segmentation_module, data_module)
+
+    # Create a DataLoader for the prediction data
+    predict_dataloader = DataLoader(DummyDataset(10), batch_size=4)
+
+    trainer.predict(
+        segmentation_module,
+        ckpt_path="checkpoints/lightning_logs/version_0/checkpoints/epoch=1-step=50.ckpt",
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:176]: I believe checkpoint file is only in your local machine. You can either mock checkpoint usage or upload a file with checkpoint, or come up with other idea.

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 2 above)
Comment (sham1lk) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:71]: I would say here: tests/integration/models/torch/lighting_modules/conftest.py

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 3 above)
Comment (QazyBi) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:152]: https://innofw.readthedocs.io/en/latest/contributing/ step 4.

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    dataset = DummyDataset(num_samples=100)
+    dataloader = DataLoader(dataset, batch_size=4)
+
+    # Fit the model
+    trainer.fit(segmentation_module, dataloader)
+
+    # Test the loaded model
+    trainer.test(
+        segmentation_module,
+        ckpt_path="checkpoints/lightning_logs/version_0/checkpoints/epoch=1-step=50.ckpt",
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:149]: I believe checkpoint file is only in your local machine. You can either mock checkpoint usage or upload a file with checkpoint, or come up with other idea.

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    checkpoint_callback = ModelCheckpoint(
+        dirpath=checkpoint_dir,
+        filename="segmentation_module-{epoch:02d}",
+        save_top_k=1,
+        monitor="loss",
+        mode="min",
+        save_weights_only=True,
+    )
+
+    trainer = Trainer(max_epochs=1, devices=1, callbacks=[checkpoint_callback])
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:103]: you can also add a test that runs on gpu if machine has cuda available. Check other tests for such functionality

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        return DataLoader(self.dataset, batch_size=self.batch_size)
+
+    def val_dataloader(self):
+        return DataLoader(self.dataset, batch_size=self.batch_size)
+
+    def test_dataloader(self):
+        return DataLoader(self.dataset, batch_size=self.batch_size)
+
+
+def predict(model, dataloader):
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:59]: I believe this function is redundant

**Review by QazyBi — CHANGES_REQUESTED** [review]

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 8 above)
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:59]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 7 above)
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:103]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 6 above)
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:149]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 4 above)
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:126]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    optimizer_cfg = DictConfig(fixt_optimizers.adam_optim_w_target)
+
+    module = SemanticSegmentationLightningModule(
+        model=model, losses=losses, optimizer_cfg=optimizer_cfg
+    )
+
+    return module
+
+
+def test_training_with_checkpoint(segmentation_module: LightningModule):
```
[the remark this answers is not in the document]
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:90]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    dataset = DummyDataset(num_samples=100)
+    dataloader = DataLoader(dataset, batch_size=4)
+    trainer.fit(segmentation_module, train_dataloaders=dataloader)
+
+
+def test_testing_without_checkpoint(segmentation_module: LightningModule):
+    trainer = Trainer(max_epochs=1, devices=1)
+    dataset = DummyDataset(num_samples=100)
+    dataloader = DataLoader(dataset, batch_size=4)
+    trainer.test(segmentation_module, dataloaders=dataloader)
```
[the remark this answers is not in the document]
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:120]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    dataset = DummyDataset(num_samples=100)
+    dataloader = DataLoader(dataset, batch_size=4)
+    trainer.fit(segmentation_module, train_dataloaders=dataloader)
+
+
+def test_training_without_checkpoint(segmentation_module: LightningModule):
+    trainer = Trainer(max_epochs=1, devices=1)
+    dataset = DummyDataset(num_samples=100)
+    dataloader = DataLoader(dataset, batch_size=4)
+    trainer.fit(segmentation_module, train_dataloaders=dataloader)
```
[the remark this answers is not in the document]
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:113]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 2 above)
Comment (Kamil-Sabbagh) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:71]: fixed

File: tests/integration/models/torch/lighting_modules/test_segmentation.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (QazyBi) [reply tests/integration/models/torch/lighting_modules/test_segmentation.py:69]: I don't think having a list of metrics values will be suitable here. Measure metrics then train a model for 10 epochs and compute metrics once more. Then compare metrics values and decide if it improved or not.

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        "test_BinaryRecall",
+    }
+    for metric in expected_metrics:
+        assert metric in test_results[0]
+
+
+def test_predicting_without_checkpoint(segmentation_module: LightningModule):
+    trainer = Trainer(max_epochs=1, devices=1)
+    predict_dataloader = DataLoader(DummyDataset(10), batch_size=4)
+    # predictions = predict(segmentation_module, predict_dataloader)
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:107]: remove commented lines

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    trainer.predict(segmentation_module, dataloaders=predict_dataloader)
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:108]: test that output of the predict is not empty list

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+
+def test_testing_with_checkpoint(segmentation_module: LightningModule):
+    checkpoint_path = "checkpoints"
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:112]: better to use [tmp_path](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html)

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+
+    # Test the loaded model
+    trainer.test(
+        segmentation_module,
+        ckpt_path="checkpoints/lightning_logs/version_0/checkpoints/epoch=1-step=50.ckpt",
+        dataloaders=dataloader,
+    )
+
+    # Delete the checkpoint folder
+    shutil.rmtree(checkpoint_path)
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:135]: if you use tmp_path you don't have to delete folder manually

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+
+
+def test_predicting_with_checkpoint(segmentation_module: LightningModule):
+    checkpoint_path = "checkpoints"
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:139]: same about path

File: tests/integration/models/torch/lighting_modules/test_segmentation.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+    # Fit the model (train and validate)
+    trainer.fit(segmentation_module, data_module)
+
+    # Create a DataLoader for the prediction data
+    predict_dataloader = DataLoader(DummyDataset(10), batch_size=4)
+
+    trainer.predict(
+        segmentation_module,
+        ckpt_path="checkpoints/lightning_logs/version_0/checkpoints/epoch=1-step=50.ckpt",
```
Comment (QazyBi) [inline tests/integration/models/torch/lighting_modules/test_segmentation.py:157]: seems like you're training a model twice to obtain similar checkpoints. I would suggest using fixture with scope session:
def get_model_ckpt(tmp_path):
    temp_dir = tmpdir_factory.mktemp('temp') 
    # use temp_dir  in trainer ...
    trainer.fit(model, datamodule)
    yield temp_dir / "checkpoints/lightning_logs/version_0/checkpoints/epoch=1-step=50.ckpt"
    

For more resources: 
https://habr.com/ru/articles/448786/
https://betterprogramming.pub/understand-5-scopes-of-pytest-fixtures-1b607b5c19ed?gi=487ecefbaf64

**Review by QazyBi — CHANGES_REQUESTED** [review]

File: innofw/core/models/torch/lightning_modules/segmentation.py [hunk 18]
```diff
@@ -133,6 +134,7 @@ def stage_step(self, stage, batch, do_logging=False, *args, **kwargs):
         #     predictions = torch.sigmoid(predictions)
 
         output[SegOutKeys.predictions] = predictions
+        self.training_metric_values.append(self.train_metrics.clone())
```
Comment (QazyBi) [inline innofw/core/models/torch/lightning_modules/segmentation.py:137]: self.training_metric_values is not needed and it occupies extra RAM. Better to remove it

**Pull request merged**